src/compiler/checker.ts

```diff
@@ -3426,12 +3515,18 @@ module ts {
                         else {
                             type = getDeclaredTypeOfSymbol(symbol);
                             if (type.flags & (TypeFlags.Class | TypeFlags.Interface) && type.flags & TypeFlags.Reference) {
-                                let typeParameters = (<InterfaceType>type).typeParameters;
-                                if (node.typeArguments && node.typeArguments.length === typeParameters.length) {
-                                    type = createTypeReference(<GenericType>type, map(node.typeArguments, getTypeFromTypeNode));
+                                // In a type reference, the outer type parameters of the referenced class or interface are automatically
+                                // supplied as type arguments and the type reference only specifies arguments for the local type parameters
+                                // of the class or interface.
+                                let localTypeParameters = (<InterfaceType>type).localTypeParameters;
+                                let expectedTypeArgCount = localTypeParameters ? localTypeParameters.length : 0;
+                                let typeArgCount = node.typeArguments ? node.typeArguments.length : 0;
+                                if (typeArgCount === expectedTypeArgCount) {
+                                    type = createTypeReference(<GenericType>type, concatenate((<InterfaceType>type).outerTypeParameters,

```

Comment:
By the way, I think there is an optimization you can do here. If typeArgCount and expectedTypeArgCount are zero, then `type` is already the correct thing because all the type arguments are exactly the same as the type arguments of the declared type. createTypeReference will just look up the type in the instantiations cache and find the type itself.